libs/@local/hashql/core/src/type/environment.rs

```diff
@@ -0,0 +1,370 @@
+        self.variance = match (old_variance, variance) {
+            // When going from covariant to contravariant context or vice versa, flip to
+            // contravariant
+            (Variance::Covariant, Variance::Contravariant)
+            | (Variance::Contravariant, Variance::Covariant) => Variance::Contravariant,
+
+            // When either context is invariant, the result is invariant
+            (Variance::Invariant, _) | (_, Variance::Invariant) => Variance::Invariant,
+
+            // Otherwise preserve the context
+            _ => variance,
+        };
+
+        let result = closure(self);
+        self.variance = old_variance;
+        result
+    }
+
+    pub(crate) fn in_contravariant<T>(&mut self, closure: impl FnOnce(&mut Self) -> T) -> T {
+        self.with_variance(Variance::Contravariant, closure)
+    }
+
+    pub(crate) fn in_covariant<T>(&mut self, closure: impl FnOnce(&mut Self) -> T) -> T {
+        self.with_variance(Variance::Covariant, closure)
+    }
+
+    pub(crate) fn in_invariant<T>(&mut self, closure: impl FnOnce(&mut Self) -> T) -> T {
+        self.with_variance(Variance::Invariant, closure)
+    }
+
+    pub(crate) fn in_transaction(&mut self, closure: impl FnOnce(&mut Self) -> bool) {
+        let (checkpoint, length, fatal) = self.begin_transaction();
+
+        let result = closure(self);
+
+        self.end_transaction(result, checkpoint, length, fatal);
+    }
+}
+
+// We usually try to avoid `Deref` and `DerefMut`, but it makes sense in this case.
+// As the unification environment is just a wrapper around the environment with an additional guard.
+impl Deref for UnificationEnvironment<'_> {
+    type Target = Environment;
+
+    fn deref(&self) -> &Self::Target {
+        self.environment
+    }
+}
+
+impl DerefMut for UnificationEnvironment<'_> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.environment
+    }
+}
+
+pub struct StructuralEquivalenceEnvironment<'env> {
+    environment: &'env Environment,
+    boundary: RecursionBoundary,
+}
+
+impl<'env> StructuralEquivalenceEnvironment<'env> {
+    #[must_use]
+    pub fn new(environment: &'env Environment) -> Self {
+        Self {
+            environment,
+            boundary: RecursionBoundary::new(),
+        }
+    }
+
+    /// Determines if two types are structurally equivalent - meaning they have the same shape and
+    /// matching internal types.
+    ///
+    /// For example:
+    /// - Two structs with the same field names and types are structurally equivalent
+    /// - Two closures with the same parameter types and return type are structurally equivalent
+    /// - Two generic types are structurally equivalent if their parameters and constraints match
+    ///
+    /// This function handles recursive types by using a recursion guard to prevent infinite
+    /// recursion.
+    ///
+    /// # Returns
+    ///
+    /// `true` if the types are structurally equivalent, `false` otherwise
```

Comment:
A quite redundant comment I think

Reply:
removed in [a4a7829](https://github.com/hashintel/hash/pull/6918/commits/a4a7829ca46e5f4a762eb74fe98f9b706c2e2780)